### PR TITLE
Documentation: Matching Utilities - isAnyOf, isAllOf

### PR DESCRIPTION
**Pull request by douglas-treadwell**
This is a documentation only PR, for the `matchers` feature branch.

This adds a `Matchers` section under `API Reference` in the sidebar, with `isAllOf` and `isAnyOf` for now.

The documentation shows some simple examples.  More complex use cases may be specific to the user, but if there are some common ones I'd be happy to add them.

Let me know what you think.  I'll be happy to make improvements.

### CODE REVIEW

**Comment by markerikson** [comment]
My first thought is that these are similar enough we really only need one docs page.  Maybe call it "Matching Utilities" or something along those lines?  After all, the only real difference between these functions is the boolean operation being performed.

**Comment by douglas-treadwell** [comment]
@markerikson That sounds good.

What would you suggest after this expands to include more functions like `isAsyncThunkAction`, `isFulfilled`, `isRejected`, `isPending` and `isRejectedWithValue`?

Perhaps one item under `Matchers` called `isAnyOf / isAllOf` (or alternatively, `Simple Matchers`), and another called `Async Matchers`?

**Comment by douglas-treadwell** [comment]
Do you have any thoughts about the content of the documentation?  Are the examples sufficient?

File: docs/api/isAllOf.mdx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+          ...state,
+          isInteresting: true,
+        }
+      }
+    });
+});
+```
+
+With isAllOf, we can use matchers to reorganize that code into a more canonical form.
+And more importantly, it enables more complex cases to be handled straightforwardly.
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:84]:
```suggestion
With `isAllOf`, we can use matchers to drastically simplify the `loadingReducer`.
```

File: docs/api/isAllOf.mdx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  isSpecial: false,
+  isInteresting: false,
+}
+
+const request = createAsyncThunk('request', () => {
+  return {
+    isSpecial: true,
+    isInteresting: true,
+  };
+});
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:125]:
```suggestion
// ... omitted
```

File: docs/api/isAllOf.mdx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+const loadingReducer = createReducer(initialState, builder => {
+  builder
+    .addMatcher(isAllOf(request.fulfilled, isSpecial), (state, action) => {
+      return {
+        ...state,
+        isSpecial: true,
+      }
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:133]:
```suggestion
      state.isSpecial = true;
```

File: docs/api/isAllOf.mdx [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    })
+    .addMatcher(isAllOf(request.fulfilled, isInteresting), (state, action) => {
+      return {
+        ...state,
+        isInteresting: true,
+      }
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:139]:
```suggestion
      state.isInteresting = true;
```

File: docs/api/isAllOf.mdx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    });
+})
+```
+
+## As a TypeScript TypeGuard
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:144]:
```suggestion
## As a TypeScript Type Guard
```

File: docs/api/isAnyOf.mdx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    .addCase(request.pending, (state, action) => ({
+      isLoading: true
+    }))
+    .addMatcher(requestFinished, (state, action) => ({
+      isLoading: false
+    }))
+})
+```
+
+## As a TypeScript TypeGuard
```
Comment (msutkowski) [inline docs/api/isAnyOf.mdx:75]:
```suggestion
## As a TypeScript Type Guard
```

File: docs/api/isAllOf.mdx [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  isInteresting: true;
+}
+
+function isSpecial(action: PayloadAction<Data>): action is PayloadAction<Special> {
+  return action.payload.isSpecial;
+}
+
+function isInteresting(action: PayloadAction<Data>): action is PayloadAction<Interesting> {
+  return action.payload.isInteresting;
+}
```
Comment (msutkowski) [inline docs/api/isAllOf.mdx:171]: I think once you merge these into one page, there will be no need to show these same types & type guards this many times. I feel like the bloat takes away from where the focus should be below.

File: docs/api/isAnyOf.mdx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    .addCase(request.pending, (state, action) => ({
+      isLoading: true
+    }))
+    .addCase(request.fulfilled, (state, action) => ({
+      isLoading: false
+    }))
+    .addCase(request.rejected, (state, action) => ({
+      isLoading: false
+    }))
+})
```
Comment (msutkowski) [inline docs/api/isAnyOf.mdx:46]: I would consider keeping your `return` uniform throughout the examples - some folks are definitely going to get tripped up between implicit and explicit returns as well as mutable vs immutable updates. Being that we push `immer` so much, I'd recommend we just show:

```js
(state, action) => {
   state.isLoading = false;
}
```

**Review by msutkowski — COMMENTED** [review]
Just some small feedback. As always, I'd wait on Mark to make any changes :)

**Comment by douglas-treadwell** [comment]
> Just some small feedback. As always, I'd wait on Mark to make any changes :)

Thanks for the review @msutkowski!

File: docs/api/isAnyOf.mdx (hunk 8 above)
Comment (douglas-treadwell) [reply docs/api/isAnyOf.mdx:46]: Good idea.  Thanks!

File: docs/api/matching-utilities.mdx [hunk 9]
```diff
@@ -7,11 +7,21 @@ hide_title: true
 
 # Matching Utilities
 
-Redux Toolkit exports a couple of type-safe utility matchers that you can leverage to simplify your code.
+Redux Toolkit several type-safe utility matchers that you can leverage to simplify your code.
```
Comment (phryneas) [inline docs/api/matching-utilities.mdx:10]:
```suggestion
Redux Toolkit exports several type-safe utility matchers that you can leverage to simplify your code.
```

File: docs/api/matching-utilities.mdx [hunk 10]
```diff
@@ -7,11 +7,21 @@ hide_title: true
...
+
+### General Purpose
 
 - [`isAllOf`](#isallof) - returns true when **all** conditions are met
 - [`isAnyOf`](#isanyof) - returns true when at least one of the conditions are met
 
+### `createAsyncThunk`-specific matchers
```
Comment (phryneas) [inline docs/api/matching-utilities.mdx:17]:
```suggestion
### `createAsyncThunk`-specific matchers

All these matchers can either be called with one or more thunks as arguments, in which case they will return a matcher function for that condition and thunks, or with one actions, in which case they will match for any thunk action with said condition.
```

File: docs/api/matching-utilities.mdx [hunk 11]
```diff
@@ -29,6 +39,67 @@ It will return a type guard function that returns `true` if _all_ of the provide
...
+A higher-order function that returns a type guard function that may be used to check whether an action was created by [`createAsyncThunk`](./createAsyncThunk).
+
+```ts title="isAsyncThunkAction usage"
+import { isAsyncThunkAction } from '@reduxjs/toolkit'
+import { request, request2 } from '@virtual/matchers'
+
+if (isAsyncThunkAction(request, request2)) {
+  // These are all async thunk actions, handle the case.
+}
+```
```
Comment (phryneas) [inline docs/api/matching-utilities.mdx:53]:
```suggestion
```ts title="isAsyncThunkAction usage"
import { isAsyncThunkAction } from '@reduxjs/toolkit'
import { requestThunk1, requestThunk2 } from '@virtual/matchers'

const isARequestAction = isAsyncThunkAction(requestThunk1, requestThunk2);

function handleRequestAction(action: AnyAction) {
  if (isARequestAction(action)) {
    // action is either an action dispatched by `requestThunk1` or `requestThunk2`
  }
}
```
```

Comment (phryneas) [reply docs/api/matching-utilities.mdx:53]: the same pattern would apply to the others.

**Comment by phryneas** [comment]
Other usage examples would be
```js
if (isRejectedAction(action)) {
  // ...
}
```

and
```js
extraReducers: builder => {
  builder.addMatcher(isFulfilledAction(requestThunk1, requestThunk2), action => {
    // ...
  })
}
```

File: docs/api/matching-utilities.mdx [hunk 12]
```diff
@@ -0,0 +1,229 @@
+---
+id: matching-utilities
+title: Matching Utilities
+sidebar_label: Matching Utilities
+hide_title: true
+---
+
+# Matching Utilities
+
+Redux Toolkit exports several type-safe utility matchers that you can leverage to simplify your code.
```
Comment (markerikson) [inline docs/api/matching-utilities.mdx:10]:
```suggestion
Redux Toolkit exports several type-safe action matching utilities that you can leverage when checking for specific kinds of actions.  These are primarily useful for the `builder.addMatcher()` cases in `createSlice` and `createReducer`, as well as when writing custom middleware.
```

File: docs/api/matching-utilities.mdx [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+### General Purpose
+
+- [`isAllOf`](#isallof) - returns true when **all** conditions are met
+- [`isAnyOf`](#isanyof) - returns true when at least one of the conditions are met
```
Comment (markerikson) [inline docs/api/matching-utilities.mdx:15]:
```suggestion
- [`isAnyOf`](#isanyof) - returns true when **at least one of** the conditions are met
```

**Review by markerikson — CHANGES_REQUESTED** [review]

File: docs/api/matching-utilities.mdx [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    }
+  }
+)
+
+// This has unnecessary complexity
+const loadingReducer = createReducer(initialState, builder => {
+  builder.addCase(isSpecialAndInterestingThunk.fulfilled, (state, action) => {
+    if (isSpecial(action)) {
+      state.isSpecial = true
+    } else if (isInteresting(action)) {
```
Comment (phryneas) [inline docs/api/matching-utilities.mdx:197]: Due to the `else` here, both examples actually behave differently.
In the matcher case, both execute.

Comment (msutkowski) [reply docs/api/matching-utilities.mdx]: Updated. I just copy and pasted that from the original and didn't bother to check 🤦

**Pull request merged**